### PR TITLE
ImageAnimatorWCS

### PR DESCRIPTION
**Pull request by abit2**
subclassing the ImageAnimator that uses WCS to animate

### CODE REVIEW

**Comment by abit2** [comment]
@Cadair @DanRyanIrish @ehsteve

File: sunpy/visualization/imageanimator.py [hunk 1]
```diff
@@ -878,3 +880,29 @@ def update_plot(self, val, line, slider):
         if val != slider.cval:
             line.set_ydata(self.data[self.frame_slice])
             slider.cval = val
+
+class ImageAnimatorWCS(ImageAnimator):
+    """
+    Same parameters as ImageAnimator with a extra parameter for WCS.
+    """
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Can you just and paste the ImageAnimator parameters and add the wcs parameter?  So the docstring stands alone.

File: sunpy/visualization/imageanimator.py [hunk 2]
```diff
@@ -9,12 +9,14 @@
 import matplotlib.widgets as widgets
 import matplotlib.animation as mplanim
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Delete blank line

File: sunpy/visualization/imageanimator.py [hunk 3]
```diff
@@ -9,12 +9,14 @@
...
 
+import astropy.units as u
+
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Delete blank line

File: sunpy/visualization/imageanimator.py [hunk 4]
```diff
@@ -878,3 +880,29 @@ def update_plot(self, val, line, slider):
...
+class ImageAnimatorWCS(ImageAnimator):
+    """
+    Same parameters as ImageAnimator with a extra parameter for WCS.
+    """
+    def __init__(self, data, wcs=None, image_axes=[-2, -1], axis_ranges=None, **kwargs):
+        self.wcs = wcs
+        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+
+    def _get_main_axes(self):
+        axes = self.fig.add_axes([0.1, 0.1, 0.8, 0.8], projection=self.wcs, slices=(0, 'y', 'x'))
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: I think the ordering of 'y' and 'x' in the slices kwarg should be related from image_axes or self.image_axes?

File: sunpy/visualization/imageanimator.py [hunk 5]
```diff
@@ -878,3 +880,29 @@ def update_plot(self, val, line, slider):
...
+        axes.coords[2].set_ticks(exclude_overlapping=True)
+        axes.coords[2].set_format_unit(u.nm)
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: The units shouldn't be hard coded, but rather a user controlled kwarg.

File: sunpy/visualization/imageanimator.py [hunk 6]
```diff
@@ -878,3 +880,29 @@ def update_plot(self, val, line, slider):
...
+    """
+    Same parameters as ImageAnimator with a extra parameter for WCS.
+    """
+    def __init__(self, data, wcs=None, image_axes=[-2, -1], axis_ranges=None, **kwargs):
+        self.wcs = wcs
+        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+
+    def _get_main_axes(self):
+        axes = self.fig.add_axes([0.1, 0.1, 0.8, 0.8], projection=self.wcs, slices=(0, 'y', 'x'))
+        axes.coords[2].set_ticks(exclude_overlapping=True)
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Same comment as above but relating to whether coords[2] or another element of coords.

File: sunpy/visualization/imageanimator.py [hunk 7]
```diff
@@ -878,3 +880,29 @@ def update_plot(self, val, line, slider):
...
+
+class ImageAnimatorWCS(ImageAnimator):
+    """
+    Same parameters as ImageAnimator with a extra parameter for WCS.
+    """
+    def __init__(self, data, wcs=None, image_axes=[-2, -1], axis_ranges=None, **kwargs):
+        self.wcs = wcs
+        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+
+    def _get_main_axes(self):
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Out of interest, where is this used?

**Comment by DanRyanIrish** [comment]
When you're making your next commit, could you also move line 4: ```from sunpy.extern import six``` to just above line 15 but keeping a blank line above it?  This was an oversight of mine in my PR.

**Comment by Cadair** [comment]
For each step of the animation you need to reset the wcs and the slice to handle all cases: http://docs.astropy.org/en/stable/api/astropy.visualization.wcsaxes.WCSAxes.html#astropy.visualization.wcsaxes.WCSAxes.reset_wcs

File: sunpy/visualization/imageanimator.py (hunk 7 above)
Comment (abit2) [inline sunpy/visualization/imageanimator.py]: In _make_axis_grid function in BaseFuncAnimator

Comment (abit2) [inline sunpy/visualization/imageanimator.py]: it returns the WCSAxes

File: sunpy/visualization/imageanimator.py (hunk 4 above)
Comment (abit2) [inline sunpy/visualization/imageanimator.py]: I was thinking of taking the tuple as input from kwargs. @DanRyanIrish

File: sunpy/visualization/imageanimator.py [hunk 8]
```diff
@@ -1,19 +1,16 @@
 # -*- coding: utf-8 -*-
 
 import abc
-from sunpy.extern import six
 
 import numpy as np
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Delete blank line.

File: sunpy/visualization/imageanimator.py [hunk 9]
```diff
@@ -1,19 +1,16 @@
...
 
 import matplotlib.pyplot as plt
 import matplotlib.widgets as widgets
 import matplotlib.animation as mplanim
-
-import astropy.units as u
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: I think you still need to import astropy.units.  My previous comment was just to delete the blank line above and below.

File: sunpy/visualization/imageanimator.py [hunk 10]
```diff
@@ -905,4 +967,14 @@ def plot_start_image(self, ax):
...
- 
\ No newline at end of file
+
+    def update_plot(self, val, im, slider):
+        """Updates plot based on slider/array dimension being iterated."""
+        val = int(val)
+        ax_ind = self.slider_axes[slider.slider_ind]
+        ind = np.argmin(np.abs(self.axis_ranges[ax_ind] - val))
+        self.frame_slice[ax_ind] = ind
+        self.axes.reset_wcs(wcs=self.wcs, slices=self.slices_wcsaxes)
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Assuming the default value of ```self.slices_wcsaxes is used (```slices_wcsaxes=(0, 'y', 'x')```) does this mean the wcsaxes of the first exposure is used for all exposures?

File: sunpy/visualization/imageanimator.py [hunk 11]
```diff
@@ -878,3 +875,112 @@ def update_plot(self, val, line, slider):
...
+
+    unit_y_axis: astropy.units
+        The unit of y axis.
+
+    slices_wcsaxes: tuple
+        Slices parameter of WCSAxes. Example: (0, 'y', 'x') for 3D Data
+    Extra keywords are passed to imshow.
+
+    """
+    def __init__(self, data, wcs=None, image_axes=[-2, -1], unit_x_axis=None, unit_y_axis=None, slices_wcsaxes=None, axis_ranges=None, **kwargs):
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Do we need a slices_wcsaxes kwarg?  Surely that be inferred from the wcs and the image_axes values.

Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: I think so, as long as image animator can handle having the ordering of `image_axes` backwards. I.e. the slice can always be `'x', 'y'` but the `image_axes` can be `[-1, -2]`.

Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Right.  So the first image_axes value is the x axis and the second is the y axis?  Then you just find the corresponding axes in the wcs object.

Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: yeah

**Comment by abit2** [comment]
Created a Subclass of ImageAnimator added new parameters unit_x_axis and unit_y_axis (which is by default None that implies use the WCS default values) which helps to change units in respective axis.

Used the wcs.naxis to create a list. Image_axis is used to set the 0th index as x-axis and 1st index as y-axis in the self.slices_wcsaxes in the
    >>>axes = self.fig.add_axes([0.1, 0.1, 0.8, 0.8], projection=self.wcs, slices=self.slices_wcsaxes)

Overriding the _get_main_axis method which now returns **astropy.visualization.wcsaxes.core.WCSAxes**

plot_start_image method starts the plot of image and returns matplotlib.image.AxesImage

update_plot method updates the slider value and resets the wcs slice value according to present iteration.

**Comment by DanRyanIrish** [comment]
I tried setting the x-and y-axis units as part of the animation and they only lasted for the first frame.  Can you change this so the axis initial formatting remains through the whole animation?

File: sunpy/visualization/imageanimator.py [hunk 12]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
         if val != slider.cval:
             line.set_ydata(self.data[self.frame_slice])
             slider.cval = val
+
+class ImageAnimatorWCS(ImageAnimator):
+    """
+    Create a matplotlib backend independent data explorer for 2D images.
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: docstring needs updating to talk about WCS

File: sunpy/visualization/imageanimator.py [hunk 13]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+    - 'top': change the active slider up one
+    - 'bottom': change the active slider down one
+    - 'p': play/pause active slider
+
+    This viewer can have user defined buttons added by specifying the labels
+    and functions called when those buttons are clicked as keyword arguments.
+
+    Parameters
+    ----------
+    data: ndarray
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: all the parameters in the list should be sphinx references i.e.
```
`numpy.ndarray`
```

File: sunpy/visualization/imageanimator.py [hunk 14]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+        The data to be visualized >= 2D
+    
+    wcs: sunpycube.wcs_util.WCS
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: this should be `astropy.wcs.WCS` not the sunpy cube variant.

File: sunpy/visualization/imageanimator.py [hunk 15]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+    unit_x_axis: astropy.units
+        The unit of x axis.
+
+    unit_y_axis: astropy.units
+        The unit of y axis.
+
+    Extra keywords are passed to imshow.
+
+    """
+    def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y_axis=None, axis_ranges=None, **kwargs):
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: line length?

File: sunpy/visualization/imageanimator.py [hunk 16]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+        if wcs is None:
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: instead of testing for `None` test isinstance WCS.

Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: Also test that data.ndim is equal to wcs.naxis

File: sunpy/visualization/imageanimator.py [hunk 17]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+            self._add_colorbar(im)
+        return im
+
+    def update_plot(self, val, im, slider):
+        """Updates plot based on slider/array dimension being iterated."""
+        val = int(val)
+        ax_ind = self.slider_axes[slider.slider_ind]
+        ind = np.argmin(np.abs(self.axis_ranges[ax_ind] - val))
+        self.frame_slice[ax_ind] = ind
+        list_slices_wcsaxes = list(self.slices_wcsaxes)
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: I am pretty sure you would be ok with leaving `self.slices_wcsaxes` as a list.

File: sunpy/visualization/imageanimator.py [hunk 18]
```diff
@@ -878,3 +874,110 @@ def update_plot(self, val, line, slider):
...
+        list_slices_wcsaxes[ax_ind] = val
+        self.slices_wcsaxes = tuple(list_slices_wcsaxes)
+        self.axes.reset_wcs(wcs=self.wcs, slices=self.slices_wcsaxes)
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: This line should be in the if statement below

**Comment by Cadair** [comment]
@abit2 looking good! I am glad that this was as minimal as it turned out to be, I was thinking it was going to be a massive re-write for a moment there!

File: sunpy/visualization/imageanimator.py (hunk 18 above)
Comment (abit2) [inline sunpy/visualization/imageanimator.py]: @Cadair didn't understand this part

**Comment by DanRyanIrish** [comment]
Add entry to CHANGELOG

File: sunpy/visualization/imageanimator.py [hunk 19]
```diff
@@ -878,3 +875,117 @@ def update_plot(self, val, line, slider):
         if val != slider.cval:
             line.set_ydata(self.data[self.frame_slice])
             slider.cval = val
+
+class ImageAnimatorWCS(ImageAnimator):
+    """
+    Create a matplotlib backend independent data explorer for 2D images that uses wcs data.
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: something like "for arrays with associated astropy WCS objects" is more descriptive.

File: sunpy/visualization/imageanimator.py [hunk 20]
```diff
@@ -878,3 +875,117 @@ def update_plot(self, val, line, slider):
...
+    data: `numpy.ndarray`
+        The data to be visualized >= 2D
+    
+    wcs: `astropy.wcs.WCS`
+        The wcs data.
+
+    image_axes: `list`
+        The two axes that make the image
+
+    fig: `mpl.figure`
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]:
```
`matplotlib.figure.Figure`
```

File: sunpy/visualization/imageanimator.py [hunk 21]
```diff
@@ -878,3 +875,117 @@ def update_plot(self, val, line, slider):
...
+    colorbar: `bool`
+        Plot colorbar
+
+    button_labels: `list`
+        List of strings to label buttons
+
+    button_func: `list`
+        List of functions to map to the buttons
+
+    unit_x_axis: `astropy.units`
```
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]:
```
`astropy.units.Unit`
```
also the same below

File: sunpy/visualization/imageanimator.py (hunk 17 above)
Comment (Cadair) [inline sunpy/visualization/imageanimator.py]: fixed

File: sunpy/visualization/imageanimator.py [hunk 22]
```diff
@@ -936,16 +936,16 @@ class ImageAnimatorWCS(ImageAnimator):
     Extra keywords are passed to imshow.
 
     """
-    def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y_axis=None,
-                 axis_ranges=None, **kwargs):
+    def __init__(self, data, wcs=None, image_axes=[-2, -1], unit_x_axis=None, unit_y_axis=None,
```
Comment (abit2) [inline sunpy/visualization/imageanimator.py]: Image_axes is set to [0,1] gives the wrong plt so I added slice_axes

Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: What is the purpose of slice_axes here?  Is it the dimensions of the wcs object to use in making the axes?

Comment (abit2) [inline sunpy/visualization/imageanimator.py]: yes it is used in slice of axes as setting Image_axes to [0,1] the whole thing breaks

Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: Then this doesn't need to be a kwarg.  You can calculate ```slice_axes``` from ```image_axes``` by assuming the wcs axes are reversed.  Right?

**Comment by Cadair** [comment]
@DanRyanIrish approve?

**Comment by DanRyanIrish** [comment]
👍

**Comment by Cadair** [comment]
merge once CI passes.

**Pull request merged**